images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/mqtt.py

```diff
@@ -0,0 +1,180 @@
+            return
+
+        for topic, handler in self.topic_handlers.items():
+            if topic not in self.subscribed_topics:
+                try:
+                    await self.client.subscribe(topic)
+                    self.subscribed_topics.add(topic)
+                    logger.info(f"Subscribed to {topic}")
+                except Exception as e:
+                    logger.error(f"Failed to subscribe to {topic}: {e}")
+
+    async def listen_for_messages(self):
+        try:
+            async for message in self.client.messages:
+                topic = message.topic.value
+                handler = topic_from_handler(self.topic_handlers,topic)
+                await handler(message)
+        except Exception as e:
+            logger.error(f"Error in message loop: {e}")
+            await self.handle_disconnect()
+
+    async def handle_disconnect(self):
+        if self.status == self.STATUS_CONNECTED:
+            await self.client.__aexit__(None, None, None)
+            self.status = self.STATUS_DISCONNECTED
+            logger.info("Attempting to reconnect...")
+
+    def get_subscribed_topics(self):
+        return list(self.subscribed_topics)
+
+    def register_topic_handler(self, topic: str, handler: callable):
+        if not topic in self.topic_handlers:
+            self.topic_handlers[topic] = handler
+            if self.status == self.STATUS_CONNECTED:
+                asyncio.create_task(self.client.subscribe(topic))
+
+    def remove_topic_handler(self, topic: str):
+        if topic in self.topic_handlers:
+            del self.topic_handlers[topic]
+            logger.info(f"Removed handler for topic {topic}")
+
+    async def publish(self, topic: str, payload: any):
+        await self.message_queue.put((topic, payload))
+        print(payload)
+
+    async def _outgoing(self):
+        while True:
+            topic, payload = await self.message_queue.get()
+            try:
+                await self.client.publish(topic, payload)
+            except Exception as e:
+                logger.error(f"Failed to publish message: {e}")
+                await self.message_queue.put((topic, payload))
+
+    async def dbus_MQTTStatus(self, req):
+        return {"status": self._make_status_object()}
+
+    async def dbus_MQTTPublish(self, req):
+        try:
+            topic = req.get('topic')
+            message = req.get('payload')
+            if not topic or not message:
+                return {"status": "error", "message": "Topic or message missing"}
+            await self.publish(topic, json.dumps(message).encode('utf-8'))
+            return {"status": "ok"}
+        except Exception as e:
+            logger.error(f"Failed to publish message: {str(e)}")
+            return {"status": "error", "message": str(e)}
+
+    async def dbus_MQTTSubscribe(self, req):
+        topic = req.get('topic')
+        self.register_topic_handler(topic, self.default_handler)
+        return {"status": "ok"}
```

Comment:
This definitely is an API that does not make a whole lot of sense :)

Reply:
It's one of those features that turned into a convenient debugging option so I haven't removed it yet. I don't think there's any value to users in subscribing to topics over command line. 

I do think it would be useful to have a simple interface for creating subscriptions, so that users are able to set up MQTT communication with esp32 or other devices, but this ain't it